File: python/CourseScheduleII.py
class Solution(object):
    def findOrder(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: List[int]
        """
        graph = [[] for _ in range(numCourses)]
        visited = [0 for _ in range(numCourses)]
        taken = [0 for _ in range(numCourses)]
        res = []

        for pair in prerequisites:
            course, pre = pair
            graph[course].append(pre)

        for i in range(numCourses):
            if not self.dfs(i, visited, taken, graph, res):
                return []
            
        return res
    

    def dfs(self, i, visited, taken, graph, res):
        if taken[i]:
            return True

        if visited[i]:
            return False

        visited[i] = 1

        for pre in graph[i]:
            if not self.dfs(pre, visited, taken, graph, res):
                return False

        visited[i] = 0
        taken[i] = 1
        res.append(i)

        return True

File: java/CourseScheduleII.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        List<List<Integer>> pre = new ArrayList<List<Integer>>();
        for (int i = 0; i < numCourses; i++){
            pre.add(new ArrayList<Integer>());
        }
        
        for (int i= 0; i < prerequisites.length; i++){
            pre.get(prerequisites[i][0]).add(prerequisites[i][1]);
        }
        
        boolean[] visited = new boolean[numCourses];
        boolean[] taken = new boolean[numCourses];
        List<Integer> res = new ArrayList<Integer>();
        
        for (int i = 0; i < numCourses; i++){
            if (!dfs(i, visited, taken, pre, res)){
                return new int[0];
            }
        }
        
        int[] resArray = new int[numCourses];
        
        for (int i = 0; i < numCourses; i++){
            resArray[i] = res.get(i);
        }
        
        return resArray;
    }
    
    private boolean dfs(int courseId, boolean[] visited, boolean[] taken, List<List<Integer>> pre, List<Integer> res){
        if (taken[courseId]){
            return true;
        }
        
        if (visited[courseId]){
            return false;
        }
        
        visited[courseId] = true;
        for (int i : pre.get(courseId)){
            if (!dfs(i, visited, taken, pre, res)){
                return false;
            }
        }
        
        visited[courseId] = false;
        taken[courseId] = true;
        res.add(courseId);
        
        return true;       
    }
}
